src/operator/contrib/bilinear_resize-inl.cuh

```diff
@@ -81,48 +123,53 @@ __global__ void caffe_gpu_interp2_kernel(const int n,
     if (height1 == height2 && width1 == width2) {
       const int h1 = h2;
       const int w1 = w2;
-      for (int c = 0; c < channels; ++c) {
-        const Dtype val = data1[h1][w1][c];
-        data2[h2][w2][c] = val;
-      }
+    for (int b = 0; b < batch_size; ++b) {
+    for (int c = 0; c < channels; ++c) {
+      const Dtype val = data1[b][h1][w1][c];
+      data2[b][h2][w2][c] = val;
+    }
+    }
       return;
     }
     //
-    const Acctype h1r = rheight * h2;
+    const Acctype h1r = cu_area_pixel_compute_source_index<Acctype>(
+    rheight, h2, align_corners, /*cubic=*/false);
     const int h1 = h1r;
     const int h1p = (h1 < height1 - 1) ? 1 : 0;
     const Acctype h1lambda = h1r - h1;
     const Acctype h0lambda = Acctype(1) - h1lambda;
     //
-    const Acctype w1r = rwidth * w2;
+    const Acctype w1r = cu_area_pixel_compute_source_index<Acctype>(
+    rwidth, w2, align_corners, /*cubic=*/false);
     const int w1 = w1r;
     const int w1p = (w1 < width1 - 1) ? 1 : 0;
     const Acctype w1lambda = w1r - w1;
     const Acctype w0lambda = Acctype(1) - w1lambda;
+  for (int b = 0; b < batch_size; ++b) {
     for (int c = 0; c < channels; ++c) {
-      const Acctype val = h0lambda * (w0lambda * data1[h1][w1][c]
-                            + w1lambda * data1[h1][w1+w1p][c])
-                            + h1lambda * (w0lambda * data1[h1+h1p][w1][c]
-                            + w1lambda * data1[h1+h1p][w1+w1p][c]);
-      data2[h2][w2][c] = ScalarConvert<Acctype, Dtype>::to(val);
+      const Acctype val = h0lambda * (w0lambda * data1[b][h1][w1][c]
+                        + w1lambda * data1[b][h1][w1 + w1p][c])
+                        + h1lambda * (w0lambda * data1[b][h1 + h1p][w1][c]
+                        + w1lambda * data1[b][h1 + h1p][w1 + w1p][c]);
+      data2[b][h2][w2][c] = ScalarConvert<Acctype, Dtype>::to(val);
     }
   }
+  }
 }
 
-// caffe_gpu_interp2_kernel overloading with Tensor<xpu, 4, DType>
+// caffe_gpu_interp2_kernel overloading with Tensor<xpu, 3, DType> for HWC layout
 template<typename xpu, typename Dtype, typename Acctype>
 __global__ void caffe_gpu_interp2_kernel(const int n,
-    const Acctype rheight, const Acctype rwidth,
-    const Tensor<xpu, 4, Dtype> data1,
-    Tensor<xpu, 4, Dtype> data2,
-    ImageLayout layout) {
+  const Acctype rheight, const Acctype rwidth,
+  const bool align_corners,
+  const Tensor<xpu, 3, Dtype> data1,
+  Tensor<xpu, 3, Dtype> data2) {
   int index = threadIdx.x + blockIdx.x * blockDim.x;
-  int batch_size = (layout == NHWC) ? data1.size(0) : data1.size(0);
-  int channels = (layout == NHWC) ? data1.size(3) : data1.size(1);
-  int height1 = (layout == NHWC) ? data1.size(1) : data1.size(2);
-  int width1 = (layout == NHWC) ? data1.size(2) : data1.size(3);
-  int height2 = (layout == NHWC) ? data2.size(1) : data2.size(2);
-  int width2 = (layout == NHWC) ? data2.size(2): data2.size(3);
+  const int height1 = data1.size(0);
```

Comment:
Is the layout always NCHW?

Reply:
It's HWC. NHWC/HWC layout code is for 'nd.image.resize', it only has forward implementation before.
There are NCHW implementation too. It's the really internal method of BilinearResize2D